### PR TITLE
optimize channels last for BatchNorm2d on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55864 enable BFloat16 mkldnn_convolution on both contiguous and channels last memory format
* #55584 add channels last (2d) support for mkldnn_convolution
* #51185 add channels last support for ConvTranspose2d
* #50573 add channels last support for PixelShuffle and PixelUnshuffle
* #50247 add channels last support for ChannelShuffle
* #49984 add channel last support for MaxUnpool2d
* #49821 add channels last for GroupNorm
* #49582 add channels last support for thnn_conv2d (non-dilated)
* #48920 add channels last for AdapativeMaxPool2d
* **#48919 optimize channels last for BatchNorm2d on CPU**
* #48918 add channels last support for AvgPool2d on CPU

move data indexing utils

parallel inference contiguous path

parallel inference channels last path

add dim apply

optimize update stats

add channels last support for backward

Revert "add channels last support for backward"

This reverts commit cc5e29dce44395250f8e2abf9772f0b99f4bcf3a.

Revert "optimize update stats"

This reverts commit 7cc6540701448b9cfd5833e36c745b5015ae7643.

Revert "add dim apply"

This reverts commit b043786d8ef72dee5cf85b5818fcb25028896ecd.

bug fix

add batchnorm nhwc test for cpu, including C=1 and HW=1

Differential Revision: [D25399468](https://our.internmc.facebook.com/intern/diff/D25399468)

### CODE REVIEW

**Comment by mingfeima** [comment]
use this one to replace #46234, NHWC performance provided.

* BatchNorm training forward path is paralleled but inference path is not paralleled at the moment. This explains why inference path is slower than training forward on multiple cores in `batchnorm_M1_N256_K3136_cpu_bwdall `.
* Moving channels last impl to /cpu doesn't bring any performance improvement (gcc should be able to vectorize simple a * x + b pattern pretty well) but i still did so since the overall code structure will be cleaner. For example, you can keep just one version of `batch_norm_cpu_inference_collect_linear_and_constant_terms`.
* I noticed the original impl has a special treatment for `image_size=1`, it is vectorized in this patch, so for K1 cases there are some improvement.
* Since the vectorization pattern is separately treated for `image_size=1`, I add new tests to cover these scenarios.
* All backward related kernels are properly optimized, so the training performance has significant improvement.


### internal operator benchmark
Update performance with [pytorch internal operator benchmark](https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark), machine Xeon(R) Gold 6248 CPU, 20 cores per socket, 2.5GHz. **1C** refers to single core run, **20C** refers to single socket run. jemalloc and numactrl are applied to reduce test result fluctuation.


Name | Tag | before (1C) | before (20C) | after (1C) | after (20C)
-- | -- | -- | -- | -- | --
batchnorm_M1_N256_K3136_cpu | Short | 248.093 | 252.302 | 251.495 | 22.786
batchnorm_M1_N256_K3136_cpu_bwdall | Short | 3605.876 | 404.381 | 1321.963 | 134.620
batchnorm_M1_N256_K3136_cpu_bwd1 | Short | 3600.249 | 403.502 | 1325.088 | 134.584
batchnorm_M1_N8192_K1_cpu | Long | 26.128 | 25.874 | 23.520 | 23.495
batchnorm_M1_N2048_K1_cpu | Long | 10.377 | 10.455 | 9.987 | 10.080
batchnorm_M128_N8192_K1_cpu | Long | 665.283 | 670.573 | 369.707 | 53.901
batchnorm_M128_N2048_K1_cpu | Long | 177.345 | 184.846 | 96.405 | 22.188
batchnorm_M1_N8192_K1_cpu_bwdall | Long | 16337.803 | 5278.381 | 84.187 | 136.834
batchnorm_M1_N8192_K1_cpu_bwd1 | Long | 16343.511 | 5712.935 | 84.999 | 139.300
batchnorm_M1_N2048_K1_cpu_bwdall | Long | 4098.062 | 1397.278 | 43.828 | 80.294
batchnorm_M1_N2048_K1_cpu_bwd1 | Long | 4088.836 | 1297.217 | 43.936 | 80.556
batchnorm_M128_N8192_K1_cpu_bwdall | Long | 36745.255 | 5309.559 | 1953.939 | 300.480
batchnorm_M128_N8192_K1_cpu_bwd1 | Long | 36715.947 | 5316.020 | 1959.126 | 300.013
batchnorm_M128_N2048_K1_cpu_bwdall | Long | 7132.537 | 1307.126 | 523.393 | 170.343
batchnorm_M128_N2048_K1_cpu_bwd1 | Long | 7065.365 | 1361.857 | 524.974 | 170.881

### RN50 BatchNorm2d performance
since the internal operator benchmark covers only NCHW memory format, I abstracted BatchNorm2d shapes from resnet50, tested both NCHW and NHWC memory format on 3 scenarios:

a. single core inference -  the perf is pretty much identical to before.
b. single socket inference (BS=1) - unit: ms per iteration

input sizes | before (nchw) | before (nhwc) | after (nchw) | after (nhwc)
-- | -- | -- | -- | --
[1,64,112,112] | 0.272 | 0.508 | 0.042 | 0.041
[1,64,56,56] | 0.088 | 0.146 | 0.033 | 0.031
[1,256,56,56] | 0.272 | 0.463 | 0.040 | 0.042
[1,128,56,56] | 0.150 | 0.254 | 0.036 | 0.034
[1,128,28,28] | 0.041 | 0.080 | 0.032 | 0.029
[1,512,28,28] | 0.151 | 0.245 | 0.037 | 0.037
[1,256,28,28] | 0.089 | 0.138 | 0.033 | 0.032
[1,256,14,14] | 0.028 | 0.045 | 0.029 | 0.029
[1,1024,14,14] | 0.096 | 0.148 | 0.034 | 0.034
[1,256,14,14] | 0.028 | 0.045 | 0.029 | 0.029
[1,512,14,14] | 0.048 | 0.079 | 0.032 | 0.031
[1,512,7,7] | 0.026 | 0.032 | 0.029 | 0.029
[1,2048,7,7] | 0.060 | 0.082 | 0.034 | 0.034

c. single socket training (BS=128) - unit: ms per iteration

input sizes | before (nchw) | before (nhwc) | after (nchw) | after (nhwc)
-- | -- | -- | -- | --
[128,64,112,112] | 85.081 | 431.061 | 57.147 | 39.251
[128,64,56,56] | 21.279 | 114.859 | 14.535 | 10.368
[128,256,56,56] | 80.615 | 383.638 | 58.409 | 32.715
[128,128,56,56] | 40.969 | 166.508 | 28.946 | 17.632
[128,128,28,28] | 10.062 | 40.375 | 6.731 | 4.671
[128,512,28,28] | 42.828 | 197.343 | 29.662 | 14.609
[128,256,28,28] | 20.250 | 91.205 | 14.234 | 8.688
[128,256,14,14] | 5.112 | 14.013 | 3.064 | 2.323
[128,1024,14,14] | 23.254 | 101.761 | 14.100 | 7.146
[128,256,14,14] | 5.089 | 13.930 | 3.172 | 2.326
[128,512,14,14] | 10.760 | 46.684 | 6.564 | 3.929
[128,512,7,7] | 2.944 | 5.712 | 2.119 | 1.305
[128,2048,7,7] | 13.583 | 52.279 | 9.271 | 3.502

**Review by VitalyFedyunin — DISMISSED** [review]

File: aten/src/ATen/cpu/vec256/functional.h [hunk 1]
```diff
@@ -231,4 +231,33 @@ inline void map3(
   }
 }
 
+template <typename scalar_t, typename Op>
+inline void map4(
```
Comment (VitalyFedyunin) [inline aten/src/ATen/cpu/vec256/functional.h:235]: Unused?

Comment (mingfeima) [reply aten/src/ATen/cpu/vec256/functional.h:235]: used in [here](https://github.com/pytorch/pytorch/pull/48919/files#diff-e39a21a7125ac201b766a585b57ebf8429a7ac28cd723b09930aceb198fd25b0R510), the kernel file is folded by github since it is a bit large.

Comment (VitalyFedyunin) [inline aten/src/ATen/cpu/vec256/functional.h:235]: Please add vec256 tests

Comment (mingfeima) [reply aten/src/ATen/cpu/vec256/functional.h:235]: Test case added in `aten/src/ATen/test/vec256_test_all_types.cpp`

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 2]
```diff
@@ -87,28 +105,596 @@ void batch_norm_cpu_inference_contiguous_impl(Tensor& output,
...
+    const Tensor& running_mean, const Tensor& runnning_var, bool train, double eps) {
+
+  using Vec = Vec256<scalar_t>;
+  int64_t n_batch = input.size(0);
+  int64_t n_channel = input.size(1);
+  int64_t image_size = input.numel() / n_batch / n_channel;
+
+  Tensor alpha = at::empty({n_channel}, input.options());
+  Tensor beta = at::empty({n_channel}, input.options());
+  auto alpha_data = alpha.data_ptr<scalar_t>();
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp]: Using scalar_t* would be more straight-forward. Fixed.

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 3]
```diff
@@ -87,28 +105,596 @@ void batch_norm_cpu_inference_contiguous_impl(Tensor& output,
...
+    Tensor& mean, Tensor& var_sum, const Tensor& input) {
+
+  using Vec = Vec256<scalar_t>;
+  using accscalar_t = at::acc_type<scalar_t, false>;
+  int64_t n_batch = input.size(0);
+  int64_t n_channel = input.size(1);
+  int64_t image_size = input.numel() / n_batch / n_channel;
+  int64_t N = input.numel() / n_channel;
+
+  auto input_data = input.data_ptr<scalar_t>();
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp]: Fixed!

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 4]
```diff
@@ -87,28 +105,596 @@ void batch_norm_cpu_inference_contiguous_impl(Tensor& output,
...
+  //    {NHW, C} => {max_threads, C}
+  //
+  // Second path: parallel along dim1 of the immediate buffer,
+  //    {max_threads, C} => {C}
+  //
+  // Normal size of C should fit in L1, otherwise consider blocking on C.
+  //
+  int num_threads = at::get_num_threads();
+  Tensor buffer = at::empty({num_threads, n_channel}, input.options()).zero_();
+  auto buffer_data = buffer.data_ptr<scalar_t>();
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp]: Fixed!

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 5]
```diff
@@ -87,28 +105,596 @@ void batch_norm_cpu_inference_contiguous_impl(Tensor& output,
...
+    Tensor& mean, Tensor& var_sum, const Tensor& input) {
+
+  using Vec = Vec256<scalar_t>;
+  using accscalar_t = at::acc_type<scalar_t, false>;
+  int64_t n_batch = input.size(0);
+  int64_t n_channel = input.size(1);
+  int64_t image_size = input.numel() / n_batch / n_channel;
+  int64_t N = input.numel() / n_channel;
+
+  auto input_data = input.data_ptr<scalar_t>();
```
[the remark this answers is not in the document]
Comment (mingfeima) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp]: Fixed!

**Comment by mingfeima** [comment]
#### update - AVX2 CPU benchmark result:
* **avx2 CPU**: Xeon(R) CPU E5-2680 v3 @ 2.50GHz, dual sockets, 12 cores per socket. Codename **Haswell**, [spec](https://ark.intel.com/content/www/us/en/ark/products/81908/intel-xeon-processor-e5-2680-v3-30m-cache-2-50-ghz.html)

single socket inference (bs=1)  - unit: ms per iteration

input sizes | before (nchw) | before (nhwc) | after (nchw) | after (nhwc)
-- | -- | -- | -- | --
[1,64,112,112] | 0.383 | 0.825 | 0.374 | 0.459
[1,64,56,56] | 0.098 | 0.200 | 0.099 | 0.112
[1,256,56,56] | 0.300 | 0.626 | 0.293 | 0.362
[1,128,56,56] | 0.168 | 0.330 | 0.164 | 0.192
[1,128,28,28] | 0.065 | 0.107 | 0.061 | 0.069
[1,512,28,28] | 0.173 | 0.417 | 0.167 | 0.200
[1,256,28,28] | 0.127 | 0.229 | 0.098 | 0.110
[1,256,14,14] | 0.100 | 0.096 | 0.048 | 0.049
[1,1024,14,14] | 0.145 | 0.226 | 0.113 | 0.118
[1,256,14,14] | 0.063 | 0.086 | 0.049 | 0.053
[1,512,14,14] | 0.084 | 0.108 | 0.073 | 0.074
[1,512,7,7] | 0.050 | 0.055 | 0.047 | 0.040
[1,2048,7,7] | 0.100 | 0.116 | 0.099 | 0.081


single socket training (bs=128)  - unit: ms per iteration

input sizes | before (nchw) | before (nhwc) | after (nchw) | after (nhwc)
-- | -- | -- | -- | --
[128,64,112,112] | 178.138 | 1167.151 | 112.528 | 82.865
[128,64,56,56] | 35.656 | 198.460 | 27.358 | 16.406
[128,256,56,56] | 158.050 | 972.093 | 105.378 | 56.357
[128,128,56,56] | 76.355 | 428.683 | 49.584 | 27.825
[128,128,28,28] | 20.006 | 98.652 | 13.623 | 7.732
[128,512,28,28] | 71.743 | 501.972 | 57.228 | 29.945
[128,256,28,28] | 45.306 | 220.398 | 24.792 | 14.401
[128,256,14,14] | 15.217 | 37.691 | 6.800 | 3.452
[128,1024,14,14] | 48.478 | 265.597 | 26.659 | 11.536
[128,256,14,14] | 13.286 | 32.654 | 8.880 | 3.861
[128,512,14,14] | 22.367 | 117.011 | 13.755 | 6.695
[128,512,7,7] | 7.187 | 12.027 | 5.033 | 1.713
[128,2048,7,7] | 37.791 | 112.453 | 20.328 | 4.609

**Comment by glaringlee** [comment]
@mingfeima I am going to approve this diff. Please rebase your PR one more time, since your version is based on a problematic master repo which caused all the doc build failed. I am not sure whether there will be merge conflict when landing it.

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 6]
```diff
@@ -87,28 +105,596 @@ void batch_norm_cpu_inference_contiguous_impl(Tensor& output,
...
+      }
+    }
+  });
+}
+
+template <typename scalar_t>
+void batch_norm_cpu_collect_stats_contiguous_impl(
+    Tensor& mean, Tensor& var_sum, const Tensor& input) {
+
+  using Vec = Vec256<scalar_t>;
```
Comment (glaringlee) [inline aten/src/ATen/native/cpu/batch_norm_kernel.cpp:191]: Please remove this line. This is not used by any code within the function and caused a internal test failure.

Comment (mingfeima) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp]: Fixed! Please review!

**Review by glaringlee — APPROVED** [review]
LGTM now

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D25399468).
